sys/fs/squashfs/squashfs_inode.c

```diff
@@ -0,0 +1,266 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+#include <sys/param.h>
+#include <sys/systm.h>
+#include <sys/buf.h>
+#include <sys/conf.h>
+#include <sys/fcntl.h>
+#include <sys/libkern.h>
+#include <sys/limits.h>
+#include <sys/lock.h>
+#include <sys/malloc.h>
+#include <sys/mount.h>
+#include <sys/mutex.h>
+#include <sys/namei.h>
+#include <sys/priv.h>
+#include <sys/proc.h>
+#include <sys/queue.h>
+#include <sys/sbuf.h>
+#include <sys/stat.h>
+#include <sys/uio.h>
+#include <sys/vnode.h>
+
+#include <squashfs.h>
+#include <squashfs_io.h>
+#include <squashfs_mount.h>
+#include <squashfs_block.h>
+#include <squashfs_inode.h>
+#include <squashfs_decompressor.h>
+
+/* Swapendian functions for all types of inodes */
+void		swapendian_base_inode(struct sqsh_base_inode *temp);
+void		swapendian_reg_inode(struct sqsh_reg_inode *temp);
+void		swapendian_lreg_inode(struct sqsh_lreg_inode *temp);
+
+/* init functions for all types of inodes */
+sqsh_err	sqsh_init_reg_inode(struct sqsh_mount *ump, struct sqsh_inode *inode);
+sqsh_err	sqsh_init_lreg_inode(struct sqsh_mount *ump, struct sqsh_inode *inode);
+
+void
+sqsh_metadata_run_inode(struct sqsh_block_run *cur, uint64_t id, off_t base)
+{
+	cur->block = (id >> 16) + base;
+	cur->offset = id & 0xffff;
+}
+
+enum vtype
+sqsh_inode_type(int inode_type)
+{
+	switch (inode_type) {
+		case SQUASHFS_DIR_TYPE:
+		case SQUASHFS_LDIR_TYPE:
+			return (VDIR);
+		case SQUASHFS_REG_TYPE:
+		case SQUASHFS_LREG_TYPE:
+			return (VREG);
+		case SQUASHFS_SYMLINK_TYPE:
+		case SQUASHFS_LSYMLINK_TYPE:
+			return (VLNK);
+		case SQUASHFS_BLKDEV_TYPE:
+		case SQUASHFS_LBLKDEV_TYPE:
+			return (VBLK);
+		case SQUASHFS_CHRDEV_TYPE:
+		case SQUASHFS_LCHRDEV_TYPE:
+			return (VCHR);
+		case SQUASHFS_FIFO_TYPE:
+		case SQUASHFS_LFIFO_TYPE:
+			return (VFIFO);
+		case SQUASHFS_SOCKET_TYPE:
+		case SQUASHFS_LSOCKET_TYPE:
+			return (VSOCK);
+	}
+	return (VBAD);
+}
+
+sqsh_err
+sqsh_verify_inode(struct sqsh_mount *ump, struct sqsh_inode *inode)
+{
+
+	/* check for inode type */
+	if (inode->base.inode_type < 1 || inode->base.inode_type > 14)
+		return (SQFS_ERR);
+
+	/* check for inode_number */
+	if (inode->base.inode_number < 1 || inode->base.inode_number > ump->sb.inodes)
```

Comment:
What makes this range valid or invalid? E.g., why is 0 an error?

Reply:
The inode numbers are from 1 to the total number of inodes.
Note that 0 is always invalid because we will always have at least a root inode.

Reply:
Done.